### PR TITLE
cabana: refactor message id from QString to struct

### PR DESCRIPTION
**Pull request by pd0wm**
Packing and unpacking QString to get the bus and address didn't seem like the cleanest solution. Was running into issues when experimenting with supporting virtual messages derived from can data (e.g. parsing UDS) that needed extra identifiers. With this struct in place it should be easy to add other identifying fields to `MessageId`. This can also be used for things like automotive ethernet where the identifier contains an IP address.

Switched to `std::optional<MessageId>` when empty strings were used to denote that no message id was set.

@deanlee Any thoughts? Any specific reason you started with using QString to represent the source, addr pair?

### CODE REVIEW

File: tools/cabana/messageswidget.cc [hunk 1]
```diff
@@ -106,7 +106,7 @@ QVariant MessageListModel::data(const QModelIndex &index, int role) const {
   if (role == Qt::DisplayRole) {
     switch (index.column()) {
       case 0: return msgName(id);
-      case 1: return id;
+      case 1: return id.toString(); // TODO: put source and address in separate columns
```
Comment (pd0wm) [inline tools/cabana/messageswidget.cc:109]: Noticed the "bus:addr" format is confusing for new users. But functional changes are for a separate PR.

**Comment by deanlee** [comment]
no special reason. is just for initial quick development.  In fact, I have long wanted to change it to this.

**Comment by deanlee** [comment]
how about use MessageId for `updateLastMsgsTo` too?
https://github.com/commaai/openpilot/blob/7f6085805690eace8d19068a319aea2d1fffea27/tools/cabana/streams/abstractstream.cc#L61

**Comment by pd0wm** [comment]
Good to know. Hopefully this can be merged quickly, otherwise we'll have a bunch of conflicts.

**Comment by adeebshihadeh** [comment]
Looks like this has conflicts already, must've been something I merged before seeing this. I'll get this merged as soon as they're fixed!

**Comment by adeebshihadeh** [comment]
Until we've deprecated web cabana and consider the new one stable, I'm also fine with one of you guys managing your own master branch and periodically opening a PR to sync up. Should even be easy enough to have GH Actions do it on a regular schedule.

**Comment by pd0wm** [comment]
@adeebshihadeh Just rebased!

I'm worried the PR will get too big or we get out of sync with upstream changes. If you don't mind quickly looking over the PRs I'm happy with the current situation.

**Comment by adeebshihadeh** [comment]
Don't mind at all, just an offer if that works better for you guys :)

**Comment by jon-chuang** [comment]
Hi, I'm getting e.g.
```
./tools/cabana/dbcmanager.h:33:1: error: unknown type name 'Q_DECLARE_METATYPE'
```

error after this PR. Any ideas on how to fix?

**Comment by jon-chuang** [comment]
Turns out you need `#include <QVariant>` for Qt5

**Pull request merged**